cmd/whisperImporter/whisperReader/main.go

```diff
@@ -0,0 +1,298 @@
+	return points
+}
+
+func getMetric(w *whisper.Whisper, file string) *whisperImporter.Metric {
+	if len(w.Header.Archives) == 0 {
+		throwError(fmt.Sprintf("ERROR: Whisper file contains no archives: %s", file))
+	}
+
+	chunksByArchive := make(whisperImporter.ArchivesOfChunks, len(w.Header.Archives))
+	md := getMetricData(getMetricName(file), int(w.Header.Archives[0].SecondsPerPoint))
+
+	for archiveIdx, archive := range w.Header.Archives {
+		points, err := w.DumpArchive(archiveIdx)
+		if err != nil {
+			throwError(fmt.Sprintf("ERROR: Failed to read archive %d in '%s', skipping: %s", archiveIdx, file, err))
+		}
+
+		chunks := make([]*chunk.Chunk, 0)
+		curChunk := -1
+
+		var point whisper.Point
+		var t0 uint32
+
+		for _, point = range sortPoints(points) {
+			if point.Timestamp == 0 {
+				continue
+			}
+
+			log(fmt.Sprintf("Processing point %+v", point))
+
+			t0 = point.Timestamp - (point.Timestamp % chunkSpan)
+			if len(chunks) == 0 || t0 > chunks[curChunk].T0 {
+				log(fmt.Sprintf("Create new chunk at t0: %d", t0))
+
+				chunks = append(chunks, chunk.New(t0))
+				curChunk++
+
+				if curChunk > 0 && !chunks[curChunk].Closed {
+					log(fmt.Sprintf("Mark previous chunk %d as finished", curChunk-1))
+					chunks[curChunk-1].Finish()
+				}
+			}
+
+			err := chunks[curChunk].Push(point.Timestamp, point.Value)
+			if err != nil {
+				throwError(fmt.Sprintf("ERROR: Failed to push value into chunk %d: %s", curChunk, err))
+			}
+		}
+
+		// if the last written point was also the last one of the
+		// current chunk we close the chunk
+		if point.Timestamp == t0+chunkSpan-archive.SecondsPerPoint {
+			log(fmt.Sprintf("Mark current chunk %d as finished", curChunk))
+			chunks[curChunk].Finish()
+		}
+
+		encodedChunks := make(whisperImporter.ArchiveOfChunks)
+		for _, chunk := range chunks {
+			if *skipUnfinishedChunks && !chunk.Closed {
+				continue
+			}
+			encodedChunks[chunk.T0] = whisperImporter.MetricChunk{
+				ChunkSpan: chunkSpan,
+				Bytes:     chunk.Series.Bytes(),
+			}
+		}
+		log(fmt.Sprintf("Whisper file %s archive %d gets %d chunks", file, archiveIdx, len(encodedChunks)))
+		chunksByArchive[archiveIdx] = encodedChunks
+	}
+
+	return &whisperImporter.Metric{
+		Metadata:   w.Header,
+		MetricData: *md,
+		Chunks:     chunksByArchive,
+	}
+}
+
+func getMetricData(name string, interval int) *schema.MetricData {
+	md := &schema.MetricData{
+		Name:     name,
+		Metric:   name,
+		Interval: interval,
+		Value:    0,
+		Unit:     "unknown",
+		Time:     0,
+		Mtype:    "gauge",
+		Tags:     []string{},
+		OrgId:    *orgId,
+	}
+	md.SetId()
+	return md
+}
+
+// scan a directory and return the list of whisper files relative to base
+func getFileList() []string {
+	fileList := make([]string, 0)
+
+	filepath.Walk(
+		*whisperDirectory,
+		func(path string, info os.FileInfo, err error) error {
+			if path[len(path)-4:] == ".wsp" {
+				fileList = append(fileList, path)
+			}
+			return nil
+		},
+	)
+
+	return fileList
```

Comment:
unbounded memory usage. why construct an entire slice, and then range over the slice to process one by one. you could just change getFileList to send to the chan and not construct a slice holding all elements.